sql/mdl.cc

```diff
@@ -1737,26 +1737,25 @@ MDL_lock::can_grant_lock(enum_mdl_type type_arg,
     {
       Ticket_iterator it(m_granted);
       MDL_ticket *ticket;
-      bool  wsrep_can_grant= TRUE;
+      can_grant = TRUE;
 
       /* Check that the incompatible lock belongs to some other context. */
       while ((ticket= it++))
       {
         if (ticket->get_ctx() != requestor_ctx &&
             ticket->is_incompatible_when_granted(type_arg))
         {
-	    if ((wsrep_thd_is_toi(requestor_ctx->get_thd()) ||
-		 wsrep_thd_is_applying(requestor_ctx->get_thd())) &&
-		key.mdl_namespace() == MDL_key::BACKUP)
-            {
+	  if ((wsrep_thd_is_toi(requestor_ctx->get_thd()) ||
+	       wsrep_thd_is_applying(requestor_ctx->get_thd())) &&
+	      key.mdl_namespace() == MDL_key::BACKUP)
+          {
             WSREP_DEBUG("global lock granted for BF: %lu %s",
                         thd_get_thread_id(requestor_ctx->get_thd()),
                         wsrep_thd_query(requestor_ctx->get_thd()));
-            can_grant = true;
           }
-          else if (!wsrep_grant_mdl_exception(requestor_ctx, ticket, &key))
+          else
           {
-            wsrep_can_grant= FALSE;
+            wsrep_handle_mdl_conflict(requestor_ctx, ticket, &key);
```

Comment:
I guess this branch should only be executed by wsrep thread.
Also for performance reasons non-wsrep thread should `break` immediately after it found incompatible lock.
To avoid code duplication, can we integrate this `WITH_WSREP` block into original one a few lines below?